### ISSUE TITLE
Upgrade to Phantom JS 2.0?

### ISSUE THREAD

**Issue by ldeveber** [karma-runner/karma-phantomjs-launcher#56]
Phantom JS 2 was released, are there plans to upgrade?

http://phantomjs.org/release-2.0.html

**Comment by maku** [comment]
Support for PhantomJs would be great.

**Comment by shuhei** [comment]
:+1:

**Comment by jbuiss0n** [comment]
:+1:

**Comment by gbrd** [comment]
:thumbsup:

**Comment by elexx** [comment]
:thumbsup:

**Comment by kristofsch** [comment]
:+1:

**Comment by carlgieringer** [comment]
karma-phantomjs-launcher depends on the NPM PhantomJS wrapper [phantomjs](https://github.com/Medium/phantomjs), which [might not upgrade to 2.x until 2.1](https://github.com/Medium/phantomjs/issues/288) due to some problems the PhantomJS team is having with creating working binaries for all platforms.

**Comment by stramel** [comment]
:+1:

**Comment by kellyrmilligan** [comment]
does anyone have a good workaround for function.bind not being available in older versions of phantom that will work with karma?  the mozilla polyfill won't work with react

**Comment by tschaub** [comment]
@kellyrmilligan I'm using [`karma-phantomjs-shim`](https://github.com/tschaub/karma-phantomjs-shim) with PhantomJS 1.x and React.  Let me know if it doesn't work for you (or if the docs aren't clear enough).

**Comment by Heshyo** [comment]
@kellyrmilligan I've been using the following simple shim (not tested on React). It allows me to use PouchDB.

```javascript
Function.prototype.bind = Function.prototype.bind || function (thisp) {
	var fn = this;
	return function () {
		return fn.apply(thisp, arguments);
	};
};
```

**Comment by kellyrmilligan** [comment]
@tschaub that is working perfectly, thanks!

**Comment by jperry** [comment]
:+1:

**Comment by Fost** [comment]
:+1:

**Comment by zeitiger** [comment]
:+1:

**Comment by mrzepinski** [comment]
:+1:

**Comment by allenmanning** [comment]
:+1:

**Comment by jasonslyvia** [comment]
:+1:

**Comment by OliverKK** [comment]
:+1:

**Comment by creat-or** [comment]
:+1:

**Comment by plestik** [comment]
Could everyone please stop posting things which add nothing to the discussion? I don't want to open this issue every time I get an email just to see another thumbs up. (Yes I realize this will trigger another notification).

**Comment by stramel** [comment]
:+1: @plestik comment

**Comment by just-boris** [comment]
Now I have a success story with new phantomjs. I installed phantom from [this ppa](https://launchpad.net/~tanguy-patte/+archive/ubuntu/phantomjs) and added environment variable `PHANTOMJS_BIN=phantomjs` to my tests.
Maybe it should get rid of phantomjs as dependency in `package.json` and release a new version for users which use system phantomjs like me?

**Comment by Dignifiedquire** [comment]
I totally agree with @just-boris. We should move phantomjs to be a peer dependency at most (as we might need to let people know that the launcher only works with a certain version of phantomjs) but other than that we should leave it up to the user to install the browser and figure out where it is, like we do with the other launchers.

**Comment by Dignifiedquire** [comment]
Moved to peerDeps, you can now install newer phantomjs if you want

**Comment by erykpiast** [comment]
Hmm, but how? I tried to explicitly depend on `phantomjs2` package, but it has different name obviously, so NPM installs version `phantomjs@1.9.18` anyway. Is it really the only possible approach, to install PhantomJS in the system and set env?

**Comment by Dignifiedquire** [comment]
For the time being yes that's the best way to do it. Npm@3 will stop installing peerdependencies so that in turn will remove the overhead of having two installations.

**Comment by erykpiast** [comment]
Well, ok. Thanks for the answer! I hope PhantomJS@2.1 will be released sooner than NPM@3 ;)

**Comment by nottinhill** [comment]
Interesting, could anybody please be so kind to provide steps how to get the launcher running with PhantomJS2?

**Comment by just-boris** [comment]
@nottinhill my success story:
```
brew install phantomjs
npm install karma-phantomjs-launcher just-boris/phantomjs --save-dev
```

My magic package forces to use system phantom js. So you need install it via brew, if you are using mac, for example

**Comment by nottinhill** [comment]
Thanks m8, testing now.

**Comment by nottinhill** [comment]
Strangely I see now:

    INFO [PhantomJS 1.9.8 (Mac OS X 0.0.0)]: Connected on socket 1p0emIMVjTatrVcwE7DV with id 42168014

It is working reproducably well with our blob functionality `new Blob([ data ]`!

Does this mean Phantom 2 is used but 1.9.8 displayed?

**Comment by just-boris** [comment]
@nottinhill nope, that means PhantomJS 1.9.8 is used here.
What do you see on the `phantomjs --version` ?

**Comment by nottinhill** [comment]
Yes that command is giving back `1.9.8`, strangely in my node_modules I have 2.0 installed (only).

**Comment by just-boris** [comment]
@nottinhill that guys https://github.com/medium/phantomjs put own phantomjs executable into `node_modules/.bin/` folder. You have to clean after that package.
To be sure, that you are doing right, remove whole `node_modules` and run `npm install` with new dependencies again

**Comment by nottinhill** [comment]
Will do, thanks, great help!

**Comment by just-boris** [comment]
@nottinhill yes, if you have a custom installation provided by `PHANTOMJS_BIN` var, you don't need to keep my package.
But in this case you will download phantom 1.9.8 on every `npm install` because https://github.com/medium/phantomjs do it while installing. My package allows to suppress it, an will be ok with custom phantom from `PHANTOMJS_BIN`

**Comment by nottinhill** [comment]
Great stuff, Boris! I posted a quick step-by-step for Bamboo here: https://github.com/just-boris/phantomjs/issues/1

**Comment by nottinhill** [comment]
@just-boris One more thing I noted, `brew install phantomjs` pulls in 1.9.8 and not 2.0.0. The formula phantomjs2 is not exitent. Any ideas?

**Comment by h20dragon** [comment]
I use a Mac - running OS X Yosemite with node v0.12.3.  I'm also using Karma with PhantomJS, however, I'm not using karma-phantomjs-launcher.  Here's how I get Karma and PhantomJS working:

1. Install phamtomjs using homebrew (I'm using version 2.0.0).
2. Ensure you set your environment variable as follows:

export PHANTOMJS_BIN=<Your full path to phantomjs>

Now, I can run without any issues as discussed by others (e.g. function.bind).

Hope this works for you!

**Comment by gillius** [comment]
I never figured out how to use phantomjs2 with this launcher, so I am using https://github.com/gskachkov/karma-phantomjs2-launcher, which is a fork and worked out of the box.

**Comment by gillius** [comment]
Apparently karma-phantomjs2-launcher users a different (older) version of phantomjs that doesn't work on all of the machines (gives openssl errors). So I am back on trying to make this work. I put `"karma-phantomjs-launcher": "0.2.3"` in my package.json and also `"phantomjs2": "2.0.2"`. I'm using node 5 (npm 3.3.6). So it warns about missing phantomjs peer dependency -- that's great I don't want phantom 1.9.8. But now I get `Cannot find plugin "karma-launcher-phantomjs"` even if I add it to the karma.conf. It seems like despite the peer dependency being only a warning, the plugin was not totally installed somehow.

**Comment by jodytate** [comment]
PhantomJS 2.1 was released yesterday https://github.com/ariya/phantomjs/issues/12970 (might be old news to some of us here)

**Comment by Dignifiedquire** [comment]
@jodytate thanks, yes we can finally properly upgrade this module :)

**Comment by dignifiedquire** [comment]
Fixed in 1.0.0
